tests/unit/test_hnsw_parallel.cpp

```diff
@@ -435,82 +484,90 @@ TYPED_TEST(HNSWTestParallel, parallelInsertSearch) {
 
-        auto parallel_insert = [&](int myID) {
-            for (labelType label = myID; label < n; label += n_threads / 2) {
-                completed_tasks[myID]++;
-                GenerateAndAddVector<TEST_DATA_T>(parallel_index, dim, label, label);
-            }
-        };
+    std::atomic<size_t> indexed_vectors(0);
+    // Insert the vectors we expect to search for.
+    for (labelType res_label = first_res_label; res_label <= last_res_label; res_label++) {
+        GenerateAndAddVector<data_t>(parallel_index, dim, res_label, res_label);
+        indexed_vectors++;
+    }
+    size_t n_threads = std::min(10U, FLOOR_EVEN(std::thread::hardware_concurrency()));
+    // Save the number fo tasks done by thread i in the i-th entry.
+    std::vector<size_t> completed_tasks(n_threads, 0);
 
-        TEST_DATA_T query_val = (TEST_DATA_T)n / 4;
-        std::atomic_int successful_searches(0);
-        auto parallel_knn_search = [&](int myID) {
+    auto parallel_insert = [&](int myID) {
+        for (labelType label = myID; label < n; label += n_threads / 2) {
             completed_tasks[myID]++;
-            // Make sure were still indexing in parallel to the search (at most 90% if the vectors
-            // were already indexed).
-            ASSERT_LT(VecSimIndex_IndexSize(parallel_index), 0.9 * n);
-            TEST_DATA_T query[dim];
-            GenerateVector<TEST_DATA_T>(query, dim, query_val);
+            if (label >= first_res_label && label <= last_res_label) {
+                continue; // Skip the vectors we already indexed.
+            }
+            GenerateAndAddVector<data_t>(parallel_index, dim, label, label);
+            indexed_vectors++;
+        }
+    };
+    std::atomic_int successful_searches(0);
+    size_t batch_size = n / 20;
+    auto parallel_knn_search = [&](int myID) {
+        size_t local_search_count = 0;
+        while (indexed_vectors < 0.95 * n) {
+            if (indexed_vectors < (local_search_count * batch_size)) {
+                usleep(100); // Wait for another batch of vectors to be indexed.
+                continue;
+            }
+            completed_tasks[myID]++;
+            data_t query[dim];
+            GenerateVector<data_t>(query, dim, query_val);
             auto verify_res = [&](size_t id, double score, size_t res_index) {
                 // We expect to get the results with increasing order of the distance between the
                 // res label and the query val (n/4, n/4-1, n/4+1, n/4-2, n/4+2, ...) The score is
                 // the L2 distance between the vectors that correspond the ids.
                 size_t diff_id = std::abs(int(id - query_val));
-                ASSERT_EQ(diff_id, (res_index + 1) / 2);
-                ASSERT_EQ(score, (dim * (diff_id * diff_id)));
+                if (is_multi) {
+                    ASSERT_EQ(diff_id, (res_index + 1) / 2);
+                    ASSERT_EQ(score, (dim * (diff_id * diff_id)));
+                    return;
```

Comment:
Why different behavior for multi?

Reply:
from `VecSimDebug_GetElementNeighborsInHNSWGraph` description:

> TODO: Implement the full version that supports MULTI as well.